pkg/volumeclaimbinder/persistent_volume_claim_binder.go

```diff
@@ -0,0 +1,216 @@
+				return kubeClient.PersistentVolumeClaims(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(resourceVersion string) (watch.Interface, error) {
+				return kubeClient.PersistentVolumeClaims(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), resourceVersion)
+			},
+		},
+		&api.PersistentVolumeClaim{},
+		claimStore,
+		0,
+	)
+	claimReflector.Run()
+
+	binder := &PersistentVolumeClaimBinder{
+		volumeStore: volumeStore,
+		claimStore:  claimStore,
+		client:      kubeClient,
+	}
+
+	return binder
+}
+
+func (controller *PersistentVolumeClaimBinder) Run(period time.Duration) {
+	glog.V(5).Infof("Starting PersistentVolumeClaimBinder\n")
+	go util.Forever(func() { controller.synchronize() }, period)
+}
+
+// Synchronizer is a generic List/ProcessFunc used by the Reconcile function & reconciliation loop,
+// because we're reconciling two Kinds in this component and I didn't want to dupe the loop
+type Synchronizer struct {
+	ListFunc      func() []interface{}
+	ReconcileFunc func(interface{}) error
+}
+
+func (controller *PersistentVolumeClaimBinder) synchronize() {
+	volumeSynchronizer := Synchronizer{
+		ListFunc:      controller.volumeStore.List,
+		ReconcileFunc: controller.syncPersistentVolume,
+	}
+
+	claimsSynchronizer := Synchronizer{
+		ListFunc:      controller.claimStore.List,
+		ReconcileFunc: controller.syncPersistentVolumeClaim,
+	}
+
+	controller.reconcile(volumeSynchronizer, claimsSynchronizer)
+}
+
+func (controller *PersistentVolumeClaimBinder) reconcile(synchronizers ...Synchronizer) {
+	for _, synchronizer := range synchronizers {
+		items := synchronizer.ListFunc()
+		if len(items) == 0 {
+			continue
+		}
+		wg := sync.WaitGroup{}
+		wg.Add(len(items))
+		for ix := range items {
+			go func(ix int) {
+				defer wg.Done()
+				obj := items[ix]
+				glog.V(5).Infof("Reconciliation of %v", obj)
+				err := synchronizer.ReconcileFunc(obj)
+				if err != nil {
+					glog.Errorf("Error reconciling: %v", err)
+				}
+			}(ix)
+		}
+		wg.Wait()
+	}
+}
+
+// syncPersistentVolume inspects all bound PVs to determine if their bound PersistentVolumeClaim still exists.
+func (controller *PersistentVolumeClaimBinder) syncPersistentVolume(obj interface{}) error {
+	volume := obj.(*api.PersistentVolume)
+	glog.V(5).Infof("Synchronizing persistent volume: %s\n", volume.Name)
+
+	// verify the volume is still claimed by a user
+	if volume.Spec.ClaimRef != nil {
+		if _, err := controller.client.PersistentVolumeClaims(volume.Spec.ClaimRef.Namespace).Get(volume.Spec.ClaimRef.Name); err == nil {
+			glog.V(5).Infof("PersistentVolume[%s] is bound to PersistentVolumeClaim[%s]\n", volume.Name, volume.Spec.ClaimRef.Name)
+		} else {
+			//claim was deleted by user.
+			glog.V(3).Infof("PersistentVolumeClaim[UID=%s] unbound from PersistentVolume[UID=%s]\n", volume.Spec.ClaimRef.UID, volume.UID)
+			volume.Spec.ClaimRef = nil
+			volume.Status.Phase = api.VolumeReleased
```

Comment:
Makes sense.  Leaving the ref allows recreation of the "released" status when the volume verifies its claim check no longer exists.  Future provisioning tasks can rely on this status being set at some point.